### PR TITLE
Implement `NameMatcher` in Swift

### PR DESCRIPTION
**Pull request by ahoppen**
Companion of https://github.com/apple/swift/pull/70008

---

This implements the C++ `NameMatcher` in swift-syntax, which simplifies its implementation.

### CODE REVIEW

File: Sources/SwiftIDEUtils/DeclNameLocation.swift [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+  /// A single argument inside a ``DeclNameLocation``.
+  public enum Argument: Equatable {
+    /// The argument has an explicit label.
+    ///
+    /// The ranges are the ranges of the first name, second name and colon token, each without trivia.
+    ///
+    /// ## Examples
+    /// - `a` in `func foo(a: Int) {}`
+    /// - `a` and `b` in `func foo(a b: Int) {}`
+    case labeled(firstName: Range<AbsolutePosition>, secondName: Range<AbsolutePosition>?)
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftIDEUtils/DeclNameLocation.swift:29]: The comment is out of date.

File: Sources/SwiftIDEUtils/NameMatcher.swift [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+public class NameMatcher: SyntaxAnyVisitor {
+  /// The positions that still need to be resolved.
+  ///
+  /// Elements are removed from this set when they are resoled.
+  private var positionsToResolve: [AbsolutePosition]
+
+  /// The locations that we have already resolved.
+  ///
+  /// - Note: These are not guaranteed to be in source order since we might eg. generate an entry for a subscript call
+  ///   at `[` when visiting the `SubscriptDeclSyntax` before resolving the base name at `foo[1]`.
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftIDEUtils/NameMatcher.swift:29]: You are right

File: Sources/SwiftIDEUtils/NameMatcher.swift [hunk 3]
```diff
@@ -0,0 +1,406 @@
...
+        argumentLabels: .selector(argumentNames.arguments.map { .labeled(firstName: $0.name, secondName: nil) })
+      )
+    } else if let functionCall = node.parentFunctionCall {
+      addResolvedLocIfRequested(
+        baseName: node.baseName,
+        arguments: functionCall.arguments,
+        trailingClosure: functionCall.trailingClosure,
+        additionalTrailingClosures: functionCall.additionalTrailingClosures
+      )
+    }
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftIDEUtils/NameMatcher.swift:256]: That’s handles as a `.selector` because the `DeclReferenceExprSyntax` has argument names. Good edge case though, I added a test for it.

File: Sources/SwiftIDEUtils/NameMatcher.swift [hunk 4]
```diff
@@ -0,0 +1,406 @@
...
+    for (index, positionToResolve) in positionsToResolve.enumerated() {
+      if positionToResolve > position {
+        // positionToResolve is sorted. If we're already past the position we are looking for, we won't find anything
+        // later.
+        return false
+      }
+      if positionToResolve == position {
+        positionsToResolve.remove(at: index)
+        return true
+      }
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply Sources/SwiftIDEUtils/NameMatcher.swift:78]: Good call. Feels like the sort of comment about naming I would make myself 😆 

I renamed it to `removePositionToResolveIfExists`.

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/70008

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/70008

@swift-ci Please test

**Comment by ahoppen** [comment]
https://github.com/apple/swift/pull/70008

@swift-ci Please test Windows
